bionic/flow.py

```diff
@@ -255,24 +255,14 @@ def add_case(self, case_key, names, values):
     def clear_providers(self, names):
         state = self
 
-        # TODO We delete and re-create entity defn for the given names.
-        # The recreation is vestigial of the older provider code and
-        # has a side effect of clearing names from ``default_entity_names``.
-        # Remove this logic in favor of directly clearing names.
-        original_entity_defs = [
-            state.get_entity_def(name)
-            for name in names
-            if name in state.entity_defs_by_name
-        ]
-
         # Delete the providers (or fail if not possible).
         state = state.delete_providers(names)
-        state = state.delete_entity_defs(names)
 
         # Recreate an empty version of each provider.
-        for entity_def in original_entity_defs:
-            state = state.define_entity(entity_def)
-            state = state.create_provider(entity_def.name)
+        entity_names = [name for name in names if name in state.entity_defs_by_name]
```

Comment:
nit: why not just put an if statement in the for loop?

Reply:
I thought of this too but I couldn't find a nice way to represent for and if conditional together in python. I don't think something like this is allowed:
```
for name in names if name in state.entity_defs_by_name:
  pass
```

Reply:
What I meant was:
```
for name in names:
    if name in state.entity_defs_by_name:
        ...
```